File: e2e/tests/token-management.spec.ts
import { test, expect } from '@playwright/test';
import { setup, runAuthCodeFlow } from './helpers';

test.describe('Introspection and revocation', () => {
  test('introspects an active token, revokes it and sees it die', async ({ page }) => {
    const s = await setup(page, 'tokmgmt');
    await runAuthCodeFlow(page, s);

    const accessToken = await page.inputValue('#userinfo_token');
    expect(accessToken).not.toBe('');

    // The token-management section authenticates with the selected client.
    const clientOption = page.locator('#tm_client option').filter({ hasText: s.app.name }).first();
    await page.selectOption('#tm_client', (await clientOption.getAttribute('value'))!);

    await page.locator('#introspect_token').fill(accessToken);
    await page.getByRole('button', { name: 'Introspect' }).click();
    await expect(page.locator('#introspect_result')).toContainText('"active": true');

    await page.locator('#revoke_token').fill(accessToken);
    await page.getByRole('button', { name: 'Revoke' }).click();
    await expect(page.locator('#revoke_result')).toContainText('Revoked');

    await page.getByRole('button', { name: 'Introspect' }).click();
    await expect(page.locator('#introspect_result')).toContainText('"active": false');

    // The 401 carries its error only in the WWW-Authenticate header.
    await page.getByRole('button', { name: 'Get UserInfo' }).click();
    await expect(page.locator('#userinfo_result')).toContainText('HTTP 401');
    await expect(page.locator('#userinfo_result')).toContainText('invalid_token');
  });
});

File: spec/e2e/token_management_spec.rb
# frozen_string_literal: true

require_relative "e2e_helper"

RSpec.describe "Introspection and revocation" do
  it "introspects an active token, revokes it and sees it die" do
    client = dashboard_client("tokmgmt")
    run_auth_code_flow(client)

    access_token = find("#userinfo_token").value
    expect(access_token).not_to be_empty

    # The token management section authenticates as the selected client.
    find("#tm_client option", text: client.name).select_option

    fill_in "introspect_token", with: access_token
    click_button "Introspect"
    expect(page).to have_css("#introspect_result", text: '"active": true')

    fill_in "revoke_token", with: access_token
    click_button "Revoke"
    expect(page).to have_css("#revoke_result", text: "Revoked")

    click_button "Introspect"
    expect(page).to have_css("#introspect_result", text: '"active": false')

    # The 401 carries its error only in the WWW-Authenticate header.
    click_button "Get UserInfo"
    expect(page).to have_css("#userinfo_result", text: "HTTP 401")
    expect(page).to have_css("#userinfo_result", text: "invalid_token")
  end
end
